library/blob.c

```diff
@@ -2103,46 +2139,80 @@ int eblob_plain_writev(struct eblob_backend *b, struct eblob_key *key,
 	 * TODO: We can preform read-modify-write cycle here but it's too hacky.
 	 */
 	if ((flags & BLOB_DISK_CTL_EXTHDR)
-			&& !(wc.flags & BLOB_DISK_CTL_EXTHDR)) {
+			&& !(wc->flags & BLOB_DISK_CTL_EXTHDR)) {
 		err = -ENOTSUP;
-		goto err_out_unlock;
+		goto err_out_cleanup_wc;
 	}
 
 	/*
 	 * We can only overwrite keys inplace if data-sort is not processing
 	 * this base (so binlog for it is not enabled)
 	 */
-	if (eblob_binlog_enabled(&wc.bctl->binlog)) {
+	if (eblob_binlog_enabled(&wc->bctl->binlog)) {
 		struct eblob_ram_control rctl;
 
 		err = eblob_cache_lookup(b, key, &rctl, NULL);
 		if (err != 0)
-			goto err_out_unlock;
+			goto err_out_cleanup_wc;
 
 		/* FIXME: We are possibly oversubscribing size here */
-		wc.flags = 0;
-		err = eblob_write_prepare_disk_ll(b, key, &wc,
-				wc.total_data_size + bounds.max,
+		wc->flags = 0;
+		err = eblob_write_prepare_disk_ll(b, key, wc,
+				wc->total_data_size + bounds.max,
 				EBLOB_COPY_RECORD, 0, &rctl);
 		if (err != 0)
-			goto err_out_unlock;
-		prepared = 1;
+			goto err_out_cleanup_wc;
+		*prepared = 1;
 	}
 
-	wc.flags = eblob_validate_ctl_flags(b, flags);
+	pthread_mutex_unlock(&b->lock);
+
+	wc->flags = eblob_validate_ctl_flags(b, flags);
+
+	return err;
+
+err_out_cleanup_wc:
+	eblob_write_control_cleanup(wc);
+err_out_unlock:
+	pthread_mutex_unlock(&b->lock);
```

Comment:
eblob_binlog_enabled() check is not needed in eblob_try_overwritev(), because after calling eblob_fill_write_control_from_ram() bctl is taken on hold, so binlog could not be enabled (since defrag waits on all bctl's) until bctl is not released.